### DISCUSSION TITLE
Padding in ConvTranspose

### DISCUSSION THREAD

**Discussion by goktugguvercin** [google/flax#3023] [Q&A]
While building a model with FLAX, I noticed a small detail about padding option: Setting the parameter `padding` to a numerical value such as $1$ and a string value like `"SAME"` are not same, they seem like they use different formula for the calculation of output shape. 

```
input = jax.numpy.ones((4, 8, 8, 64))
layer = nn.ConvTranspose(32, kernel_size=(3,3), strides=(2,2), padding=1)
variables = layer.init(jax.random.PRNGKey(0), input)

y = layer.apply(variables, input)
print(y.shape)
```

When I execute the code script above, output shape will be `(4, 15, 15, 32)`. In this case, the following formula is used, so the shape of height and width is equal to 15: $H_{new} = s \cdot (H - 1) - 2p + d \cdot (f - 1) + 1$

However, if we replace the value 1 in  padding argument of transpose convolution by `"SAME"` and reexecute the code again, the shape of the output will be `(4, 16, 16, 32)`. 

What is the reason lying behind it ? 
Is additionally *output padding* added to have same shape ?

**Comment by cgarciae** [comment]
Hey @GoktugGuvercin, not entirely sure exactly how SAME is implemented, my guess its that it internally results in something like setting `padding=((1, 2), (1, 2))`:

```python
layer = nn.ConvTranspose(32, kernel_size=(3,3), strides=(2, 2), padding=((1, 2), (1, 2)))
variables = layer.init(jax.random.PRNGKey(0), input)

y = layer.apply(variables, input)
print(y.shape) # (4, 16, 16, 32)
```
The intuition I have is that SAME follows this formula: $H_{new} = s \cdot H$
